flow/display_list.h

```diff
@@ -0,0 +1,427 @@
+
+  void clipRect(const SkRect& rect, bool isAA, SkClipOp clip_op) override;
+  void clipRRect(const SkRRect& rrect, bool isAA, SkClipOp clip_op) override;
+  void clipPath(const SkPath& path, bool isAA, SkClipOp clip_op) override;
+
+  void drawPaint() override;
+  void drawColor(SkColor color, SkBlendMode mode) override;
+  void drawLine(const SkPoint& p0, const SkPoint& p1) override;
+  void drawRect(const SkRect& rect) override;
+  void drawOval(const SkRect& bounds) override;
+  void drawCircle(const SkPoint& center, SkScalar radius) override;
+  void drawRRect(const SkRRect& rrect) override;
+  void drawDRRect(const SkRRect& outer, const SkRRect& inner) override;
+  void drawPath(const SkPath& path) override;
+  void drawArc(const SkRect& bounds,
+               SkScalar start,
+               SkScalar sweep,
+               bool useCenter) override;
+  void drawPoints(SkCanvas::PointMode mode,
+                  uint32_t count,
+                  const SkPoint pts[]) override;
+  void drawVertices(const sk_sp<SkVertices> vertices,
+                    SkBlendMode mode) override;
+  void drawImage(const sk_sp<SkImage> image,
+                 const SkPoint point,
+                 const SkSamplingOptions& sampling) override;
+  void drawImageRect(const sk_sp<SkImage> image,
+                     const SkRect& src,
+                     const SkRect& dst,
+                     const SkSamplingOptions& sampling) override;
+  void drawImageNine(const sk_sp<SkImage> image,
+                     const SkIRect& center,
+                     const SkRect& dst,
+                     SkFilterMode filter) override;
+  void drawImageLattice(const sk_sp<SkImage> image,
+                        const SkCanvas::Lattice& lattice,
+                        const SkRect& dst,
+                        SkFilterMode filter) override;
+  void drawAtlas(const sk_sp<SkImage> atlas,
+                 const SkRSXform xform[],
+                 const SkRect tex[],
+                 const SkColor colors[],
+                 int count,
+                 SkBlendMode mode,
+                 const SkSamplingOptions& sampling,
+                 const SkRect* cullRect) override;
+  void drawPicture(const sk_sp<SkPicture> picture) override;
+  void drawDisplayList(const sk_sp<DisplayList> display_list) override;
+  void drawTextBlob(const sk_sp<SkTextBlob> blob,
+                    SkScalar x,
+                    SkScalar y) override;
+  // void drawShadowRec(const SkPath&, const SkDrawShadowRec&) override;
+  void drawShadow(const SkPath& path,
+                  const SkColor color,
+                  const SkScalar elevation,
+                  bool occludes) override;
+
+  sk_sp<DisplayList> build();
+
+ private:
+  SkAutoTMalloc<uint8_t> storage_;
+  size_t used_ = 0;
+  size_t allocated_ = 0;
+  int opCount_ = 0;
+  int saveLevel_ = 0;
+
+  SkRect cull_;
+
+  template <typename T, typename... Args>
+  void* push(size_t extra, Args&&... args);
```

Comment:
We should really, really try to avoid `void*`. Related to my other comment above.

Reply:
Boilerplate from SkLiteDL used in both Android and Chrome. What would you suggest instead?

Reply:
Chrome uses a template here: https://source.chromium.org/chromium/chromium/src/+/main:cc/paint/paint_op_buffer.h;l=1109?q=SkLiteDL&ss=chromium%2Fchromium%2Fsrc

Reply:
They also still have some void pointers, but some of it is abstracted away or more contained and avoids _some_ pointer arithmetic in https://source.chromium.org/chromium/chromium/src/+/main:cc/paint/paint_op_reader.cc

Part of the challenge is that code like this is really hard to audit for security/safety purposes. And that will become more true as we extend/refactor it (unless we specifically refactor it for safety auditing purposes :).  If we can do any of that now we should.

But AFAICT, SkLiteDL is not actually used anywhere in Chrome or Android.

Reply:
By "using" I meant that Skia had provided SkLiteDL as a template for them to use. They customized it to their needs, but you can see the basic bones of SkLiteDL in the structure of their code. You found the Chromium variant, but Android has their own as well - similarly customized.

Reply:
Can we customize this to avoid some of the pointer arithmetic? :)

Reply:
I'm not sure what you are suggesting? Making the structs full classes? That adds overhead of a vtable per op even though the actions are only needed in one or two methods - currently those use a switch statement to inline the 2 implementations they care about (dispatch and destructor).

Since the vast majority of the entries in the list are trivially destructible the destructor can dance along the array only calling unref on the few sk_sp<> references that dot the list and then free the array. If we made them all virtual then each op would have to be manually destructed by a virtual pointer.

The comparison method can also do 99+% of its work by a memcmp in this scenario.

The memory of all of the ops is localized which would go away if we converted to some sort of virtual typesafe list of pointers to the structures. It might end up localized, but that might deteriorate over time.

Perhaps I'm not familiar with what you are suggesting, but those are the advantages that it looks like the design was aiming to accomodate.

Reply:
I like the suggestion from @chinmaygarde to move the op definitions into another file and format so that we can switch up the implementation more easily. We could then investigate the performance impact of something like a vector of virtual dispatch struct/classes.

The one question that I don't think could be easily answered from the above list of "advantages" of the current approach, though, would be the impact of memory fragmentation on locality and, by extension, on app performance. That situation would only come up in practice on long running apps. Has something like that shown up as an issue before?